### PR TITLE
[FLINK-33059] Support transparent compression for file-connector for all file input formats

### PR DESCRIPTION
**Pull request by echauchot**
## What is the purpose of the change

Support transparent compression for file-connector for all file input formats.

## Brief change log

- Force reading the whole file split (like it is done for binary input formats) on compressed (unsplittable) files
- add FileInputFormatTest#testFileInputFormatWithCompressionFromFileSource
- generalize FileInputFormatTest#testFileInputFormatWithCompression to more that deflate format

## Verifying this change
FileInputFormatTest#testFileInputFormatWithCompressionFromFileSource

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? website

### CODE REVIEW

**Comment by echauchot** [comment]
Hi @rmetzger, I saw you authored parts of this code, can you please do a review or point me to another reviewer ?

**Comment by echauchot** [comment]
R: @xintongsong I see your name in that code history, would you have time to take a look ?

**Comment by echauchot** [comment]
@tzulitai you offered help for reviewing, don't hesitate to ping me on this when you have time.

**Review by ferenc-csaky — COMMENTED** [review]
The added logic makes sense IMO.

Added 2 comments, but in general I would consider separating the whole `INFLATER_INPUT_STREAM_FACTORIES` into a new class as there are a couple functions that uses it and seems quite detachable from `FlieInputFormats`.

After a quick peek I think something like the following could work:
```java
public class InflaterInputStreamFactories {

  public static void register(String fileExt, InflaterInputStreamFactory<?> factory) { ... }

  public static InflaterInputStreamFactory<?> get(Path path) { ... }

  private static InflaterInputStreamFactory<?> get(String fileExt) { ... }

  @VisibleForTesting
  public static Set<String> getSupportedCompressionFormats() { ... }
}
```

Also, `ConcurrentHashMap` can be utilized insead of the `synchronized` block, but other than that the current logic could be moved as is now.

This probably goes beyond the current PR, but I think it worth to note it. WDYT?

File: flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java
Comment (echauchot) [reply flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java:160]: :+1: thx for pointing out

File: flink-core/src/test/java/org/apache/flink/testutils/TestFileUtils.java
Comment (echauchot) [reply flink-core/src/test/java/org/apache/flink/testutils/TestFileUtils.java:141]: Agree, better to be more generic here, thx !

**Comment by echauchot** [comment]
> The added logic makes sense IMO.
> 
> Added 2 comments, but in general I would consider separating the whole `INFLATER_INPUT_STREAM_FACTORIES` into a new class as there are a couple functions that uses it and seems quite detachable from `FlieInputFormats`.
> 
> After a quick peek I think something like the following could work:
> 
> ```java
> public class InflaterInputStreamFactories {
> 
>   public static void register(String fileExt, InflaterInputStreamFactory<?> factory) { ... }
> 
>   public static InflaterInputStreamFactory<?> get(Path path) { ... }
> 
>   private static InflaterInputStreamFactory<?> get(String fileExt) { ... }
> 
>   @VisibleForTesting
>   public static Set<String> getSupportedCompressionFormats() { ... }
> }
> ```
> 
> Also, `ConcurrentHashMap` can be utilized insead of the `synchronized` block, but other than that the current logic could be moved as is now.
> 
> This probably goes beyond the current PR, but I think it worth to note it. WDYT?

I agree with the ConcurrentHashMap suggestion. Regarding creating a class just for wrapping a map that is used only in the FileInputFormat it seems overkill to me. An anyway it is indeed outside of the scope of this filesize-fix PR.

**Comment by echauchot** [comment]
@ferenc-csaky thanks for reviewing this PR ! I have addressed your comments, do I have your LGTM when the tests pass ?

**Review by ferenc-csaky — APPROVED** [review]
LGTM, if tests pass! 👍

**Comment by ruslandanilin** [comment]
Thank you @echauchot !

**Comment by echauchot** [comment]
> Thank you @echauchot !

My pleasure ! Merging

**Pull request merged**
